### PR TITLE
longboy-proto

### PR DESCRIPTION
**Pull request by rickvanprim**
Moved generic boundary outside of `longboy-proto` and cleaned up tests a bit plus made tests try multiple permutations of `(SIZE, WINDOW_SIZE)`.

### CODE REVIEW

File: crates/longboy-proto/src/receiver.rs [hunk 1]
```diff
@@ -158,13 +109,13 @@ where
         // Advance cycles.
         loop
         {
-            let index = self.cycle % self.flags.len();
+            let index = self.cycle % Constants::receiver_flags_size::<WINDOW_SIZE>();
```
Comment (JKurzer) [inline crates/longboy-proto/src/receiver.rs:112]: Thank goodness for tab completion. Is this the final monomorphic syntax?

File: crates/longboy-proto/src/receiver.rs [hunk 2]
```diff
@@ -116,33 +65,35 @@ where
...
         {
-            let cycle_i = ((datagram_cycle + Self::MAX_CYCLE) - i) % Self::MAX_CYCLE;
+            let cycle_i =
+                ((datagram_cycle + Constants::max_cycle::<WINDOW_SIZE>()) - i) % Constants::max_cycle::<WINDOW_SIZE>();
 
             // If we're before local cycle, early out.  This is effectively checking for distance
             // being out of the buffer's size, which is only possible if before because we've
             // already adanced the local cycle to catch up, if applicable.
-            if ((cycle_i + Self::MAX_CYCLE) - self.cycle) % Self::MAX_CYCLE > self.flags.len()
+            if ((cycle_i + Constants::max_cycle::<WINDOW_SIZE>()) - self.cycle) % Constants::max_cycle::<WINDOW_SIZE>()
```
Comment (JKurzer) [inline crates/longboy-proto/src/receiver.rs:89]: Can you use a let for these constants to make this like a little more immediately legible? That gets optimized away, right?

File: crates/longboy-proto/src/receiver.rs (hunk 1 above)
Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:112]: Evidently I can do the nicer associated constant form.  That only gets slightly shorter though:
```rust
<Constants<SIZE, WINDOW_SIZE>>::RECEIVER_FLAGS_SIZE
```
I think it reads a lot better as a constant value though.

File: crates/longboy-proto/src/receiver.rs (hunk 2 above)
Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:89]: Yes, these are all constant values.  See above syntax for whether you'd prefer this be captured in a local variable somewhere.

File: crates/longboy-proto/src/receiver.rs (hunk 1 above)
Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:112]: Actually I could probably alias `type Constants = Constants<SIZE, WINDOW_SIZE>` for the type and then use the shorter form.  Originally I was trying to just define them inside but that becomes a recursive evaluation funnily enough (though better to share anyways).

Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:112]: Actually that makes `Self::Constants::<name>` which is pretty shit.  I'm going to just redefine the constants based on the other thing, aka `const DATAGRAM_SIZE: usize = <Constants<SIZE, WINDOW_SIZE>>::DATAGRAM_SIZE;` and then I can use `Self::DATAGRAM_SIZE` everywhere.

**Comment by JKurzer** [comment]
Bless.

File: crates/longboy-proto/src/receiver.rs (hunk 1 above)
Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:112]: Sigh.  That still triggers a cycle when trying to use it in function signatures.  So I'll need to still use the ugly form there 🤦‍♂️

Comment (rickvanprim) [reply crates/longboy-proto/src/receiver.rs:112]: Alright local definition it is.  There's only 1 function anyways.

**Pull request merged**